src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs

```diff
@@ -0,0 +1,112 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+using AspNet.Security.OAuth.Extensions;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.Http.Authentication;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.MailRu
+{
+    public class MailRuAuthenticationHandler : OAuthHandler<MailRuAuthenticationOptions>
+    {
+        public MailRuAuthenticationHandler([NotNull] HttpClient client)
+            : base(client)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
+        {
+            // Mail.ru REST API rules:
+            //   1. method name (users.getInfo) is passed as param
+            //   2. MD5 hash computed over all params (ordered alphabetically),
+            //        concatenated 'name1=value1name2=value2...signkey' without separators (like &)
+            //   3. MD5 Hash value passed as 'sig' param
+
+            var requestParams = "app_id=" + Options.ClientId + "&format=json&method=users.getInfo&secure=1&session_key=" + tokens.AccessToken;
```

Comment:
Please consider moving that to a separate method. You can follow the same pattern as the one we used for the Instagram provider: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs#L77